src/Utilities/Callbacks/Callbacks.jl

```diff
@@ -57,7 +225,182 @@ function wall_clock_time_per_time_step(interval, array_type, comm)
+            ca_h = DGmethods.courant(
+                advective_courant,
+                solver_config,
+                direction = HorizontalDirection(),
+            )
+            cd_v = DGmethods.courant(
+                diffusive_courant,
+                solver_config,
+                direction = VerticalDirection(),
+            )
+            cd_h = DGmethods.courant(
+                diffusive_courant,
+                solver_config,
+                direction = HorizontalDirection(),
+            )
+            simtime = ODESolvers.gettime(solver_config.solver)
+            @info @sprintf(
+                """
+Courant numbers at simtime: %8.2f, Δt = %8.2f s
+    Acoustic (vertical) Courant number    = %.2g
+    Acoustic (horizontal) Courant number  = %.2g
+    Advection (vertical) Courant number   = %.2g
+    Advection (horizontal) Courant number = %.2g
+    Diffusion (vertical) Courant number   = %.2g
+    Diffusion (horizontal) Courant number = %.2g""",
+                simtime,
+                Δt,
+                c_v,
+                c_h,
+                ca_v,
+                ca_h,
+                cd_v,
+                cd_h,
+            )
+            nothing
+        end
+        return cb_cfl
+    else
+        return nothing
+    end
 end
+
+"""
+    checkpoint(
+        checkpoint_opt,
+        checkpoint_keep_one,
+        solver_config,
+        checkpoint_dir,
+    )
+
+Return a callback function that runs at `checkpoint_opt` intervals
+and stores a simulation checkpoint into `checkpoint_dir` identified
+by a running number.
+"""
+function checkpoint(
+    checkpoint_opt,
+    checkpoint_keep_one,
+    solver_config,
+    checkpoint_dir,
+)
+    cb_constr = CB_constructor(checkpoint_opt, solver_config)
+    if cb_constr !== nothing
+        cpnum = Ref(1)
+        cb_checkpoint = cb_constr() do (init = false)
+            write_checkpoint(
+                solver_config,
+                checkpoint_dir,
+                solver_config.name,
+                solver_config.mpicomm,
+                cpnum[],
+            )
+            if checkpoint_keep_one
+                rm_checkpoint(
+                    checkpoint_dir,
+                    solver_config.name,
+                    solver_config.mpicomm,
+                    cpnum[] - 1,
+                )
+            end
+            cpnum[] += 1
+            nothing
+        end
+        return cb_checkpoint
+    else
+        return nothing
+    end
+end
+
+"""
+    CB_constructor(interval, solver_config, default)
+
+Parse the specified `interval` and return the appropriate `GenericCallbacks`
+constructor. If `interval` is "default", then use `default` as the interval.
+"""
+function CB_constructor(interval::String, solver_config, default = nothing)
+    mpicomm = solver_config.mpicomm
+    solver = solver_config.solver
+    dg = solver_config.dg
+    bl = dg.balancelaw
+    secs_per_day = day(bl.param_set)
+
+    if endswith(interval, "hours")
```

Comment:
Would it be better to use regex here, e.g.
```julia
    if (m = match(r"^([0-9]+)(hours|mins|secs|smonths|sdays|shours|steps)$",interval); m !== nothing)
        ninterval = parse(Int, m[1])
        if m[2] == "hours"
            ...
 

```

Reply:
Sim time is Float64 so it would be a bit more complicated. Not sure it's worth it.